src/cockpit/polkit.py

```diff
@@ -0,0 +1,140 @@
+        for (auth_type, subject) in identities:
+            if auth_type == 'unix-user' and 'uid' in subject and subject['uid'].value == my_uid:
+                logger.debug('Authentication subject %s matches our uid %d', subject, my_uid)
+                break
+        else:
+            logger.warning('Not supporting authentication as any of %s', identities)
+            return
+
+        user_name = pwd.getpwuid(my_uid).pw_name
+        process = await asyncio.create_subprocess_exec(HELPER_PATH, user_name, cookie,
+                                                       stdin=asyncio.subprocess.PIPE,
+                                                       stdout=asyncio.subprocess.PIPE)
+        while process.returncode is None:
+            try:
+                line = await process.stdout.readline()
+            except asyncio.CancelledError:
+                logger.debug('Cancelled authentication')
+                process.terminate()
+                break
+
+            logger.debug('Read line from helper: %s', line)
+            # usually: PAM_PROMPT_ECHO_OFF Password: \n
+            if line.startswith(b'PAM_PROMPT'):
+                prompt = line.split(b' ', 1)[1].strip().decode()
+                # Don't pass this to the UI if it's "Password:" (the usual case),
+                # so that superuser.py uses the translated default
+                if prompt.startswith('Password'):
+                    prompt = ''
+
+                try:
+                    passwd = await self.responder.do_askpass('', prompt, '')
+                    logger.debug('got PAM_PROMPT %s, do_askpass returned a password', prompt)
+                except asyncio.CancelledError:
+                    logger.debug('got PAM_PROMPT %s, do_askpass cancelled', prompt)
+                    process.terminate()
+                    break
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif line.startswith(b'PAM_TEXT_INFO'):
+                logger.info(line.decode())
+            elif line.startswith(b'PAM_ERROR'):
+                logger.warning(line.decode())
+            elif line.startswith(b'SUCCESS'):
+                logger.info('Authentication succeeded')
+                break
+            elif line.startswith(b'FAILURE'):
+                logger.warning('Authentication failed')
+                break
+            else:
+                logger.warning('Unknown line from helper, aborting: %s', line)
+                process.terminate()
+                break
+
+        res = await process.wait()
+        logger.debug('helper exited with code %i', res)
+
+    @bus.Interface.Method('', 's')
+    def cancel_authentication(self, cookie):
+        logger.warning('CancelAuthentication: %s not implemented', cookie)
```

Comment:
I'm not sure if we need this. `pkexec` doesn't seem to, and our "Cancel" UI button is implemented differently (and works).

Reply:
If you wanted to implement that, you could perhaps call `.get_current_task()` (NB: not 3.6) from the `begin_authentication()` handler and keep a dict mapping cookies to tasks, and call `.cancel()` from here.

But don't do that unless we need to...

Reply:
I added a comment about this for now. Let's not implement dead code.